File: frontend/src/api/news.ts
import { api } from "./client";

export type NewsTopic =
  | "official_update"
  | "project_progress"
  | "infrastructure"
  | "market_potential"
  | "promotion";

export interface NewsArticle {
  id: number;
  canonicalUrl: string;
  sourceId: string;
  sourceName: string;
  title: string;
  summary: string | null;
  imageUrl: string | null;
  topic: NewsTopic;
  projectNames: string[];
  publishedAt: string | null;
  fetchedAt: string;
}

interface ApiNewsArticle {
  id: number;
  canonical_url: string;
  source_id: string;
  source_name: string;
  title: string;
  summary: string | null;
  image_url: string | null;
  topic: NewsTopic;
  project_names: string[];
  published_at: string | null;
  fetched_at: string;
}

interface ApiNewsList {
  items: ApiNewsArticle[];
  total: number;
  offset: number;
  limit: number;
}

export interface NewsPageResult {
  items: NewsArticle[];
  total: number;
}

function mapArticle(row: ApiNewsArticle): NewsArticle {
  return {
    id: row.id,
    canonicalUrl: row.canonical_url,
    sourceId: row.source_id,
    sourceName: row.source_name,
    title: row.title,
    summary: row.summary,
    imageUrl: row.image_url,
    topic: row.topic,
    projectNames: row.project_names,
    publishedAt: row.published_at,
    fetchedAt: row.fetched_at,
  };
}

export async function fetchNews(params: {
  offset?: number;
  limit?: number;
  topic?: NewsTopic;
  query?: string;
}): Promise<NewsPageResult> {
  const search = new URLSearchParams();
  search.set("offset", String(params.offset ?? 0));
  search.set("limit", String(params.limit ?? 12));
  if (params.topic) search.set("topic", params.topic);
  if (params.query?.trim()) search.set("q", params.query.trim());
  const response = await api.get<ApiNewsList>(`/news?${search.toString()}`);
  return { items: response.items.map(mapArticle), total: response.total };
}

File: backend/schemas/news.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

NewsTopic = Literal[
    "official_update",
    "project_progress",
    "infrastructure",
    "market_potential",
    "promotion",
]


class NewsIngestRequest(BaseModel):
    source_id: Literal["vinhomes", "vingroup-real-estate", "vinhomes-market"]
    canonical_url: str = Field(min_length=10, max_length=3000)
    title: str = Field(min_length=5, max_length=500)
    summary: str | None = Field(default=None, max_length=2000)
    image_url: str | None = Field(default=None, max_length=3000)
    topic: NewsTopic = "official_update"
    project_names: list[str] = Field(default_factory=list, max_length=12)
    published_at: datetime | None = None

    @field_validator("title", "summary", mode="before")
    @classmethod
    def normalize_text(cls, value: object) -> object:
        if not isinstance(value, str):
            return value
        return " ".join(value.split()) or None

    @field_validator("project_names", mode="after")
    @classmethod
    def unique_projects(cls, values: list[str]) -> list[str]:
        result: list[str] = []
        seen: set[str] = set()
        for raw in values:
            value = " ".join(raw.split())[:150]
            key = value.casefold()
            if value and key not in seen:
                result.append(value)
                seen.add(key)
        return result


class NewsArticleResponse(BaseModel):
    id: int
    canonical_url: str
    source_id: str
    source_name: str
    title: str
    summary: str | None
    image_url: str | None
    topic: str
    project_names: list[str]
    published_at: datetime | None
    fetched_at: datetime

    model_config = {"from_attributes": True}


class NewsListResponse(BaseModel):
    items: list[NewsArticleResponse]
    total: int
    offset: int
    limit: int


class NewsIngestResponse(BaseModel):
    article: NewsArticleResponse
    created: bool
    content_changed: bool


class NewsRetentionResponse(BaseModel):
    archived: int
    deleted: int
